design-proposals/klip-63-pause-resume-persistent-queries.md

```diff
@@ -0,0 +1,66 @@
+# KLIP-63: PAUSE and RESUME for persistent queries 
+
+**Author**: Jim Hughes (jnh5y) | 
+**Release Target**: Unknown | 
+**Status**: In Discussion | 
+**Discussion**: 
+
+**tl;dr:** Adds the ability to pause the processing of a persistent query.
+           
+## Motivation and background
+
+Users may wish to pause a query for a number of reasons and then resume processing.  As examples, users may wish to modify data pipelines built from persistent queries or to manage computational resources.
+
+Presently to manage processing, they would have to terminate a query and recreate it.  This approach means that any state would have to be restored; the goal of this work is to add the capability to pause processing temporarily while allowing Kafka consumers and state stores to be retained.
+
+## What is in scope
+
+Pausing and resuming persistent queries and the necessary language modifications to support this are in scope.
+
+## What is not in scope
+
+We are not considering the ability to pause and resume transient queries.  Additionally, modifying the consumers' state is out of scope.  (This would allow for processing to skip records or reprocess data.)
+
+## Value/Return
+
+End users will be able to manage query processing once this feature is implemented.  For example, this will allow for better control while designing data pipelines.  Pausing an upstream query will allow users to update a downstream query or other processing and iterate through design without processing all the data in a stream.
+
+## Public APIS
+
+SQL Grammar changes:
+Add PAUSE <queryID>, PAUSE ALL, RESUME <queryID>, RESUME ALL.
```

Comment:
Isn't pause/resume ALL too wide/risky?

Reply:
ksqlDB supports TERMINATE ALL which does more damage: https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/terminate/

I'm fine with or without the ALL keyword.  I prototyped it, but it does not have to stay.

Reply:
(This is a great point for us to discuss as a group!)

Reply:
Personally, I think `ALL` is fine -- I could imagine a use case for which you want to pause all queries at the same time because they form some processing pipeline and the "final result" is consumed by some downstream application (ie, external to ksqlDB) and you want to make changes downstream.